File: Solutions/LongestCommonPrefix.cpp
// Leetcode Problem #14:https://leetcode.com/problems/longest-common-prefix/
// Source:https://leetcode.com/problems/longest-common-prefix/solution

// Solution 1: Vertical Scanning | O(m *len(LCP)) time, O(1) space where m = number of strings and LCP is the longest common prefix
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        // Loop through the indices of the first string
        // Index keeps track of the length of the LCP
        for (int index = 0; index < strs[0].length(); index++) {
            // Loop through each word.
            for (int i = 1; i < strs.size(); i++) {
                // If any prefix does not match, return the LCP
                if (strs[0][index] != strs[i][index]) return strs[0].substr(0, index);
            }
        }
            
        // If execution reaches here, the first string must be the LCP
        return strs[0];
    }
};

File: Solutions/LongestCommonPrefix.py
# LC #14: https://leetcode.com/problems/longest-common-prefix/
# Source: https://leetcode.com/problems/longest-common-prefix/solution/

# Solution 1 (Vertical Scanning) | O(S) time, O(1) space  <- S is sum of all chars in all strings
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        # LCP is the empty string if the array is empty
        if not strs:
            return ""
        
        # Loop through chars of the first string
        for i in range(len(strs[0])):
            # Loop through each string of the array
            for word in strs:
                # If any prefix does not match, return the LCP
                if i >= len(word) or word[i] != strs[0][i]:
                    return strs[0][0:i]
        
        # If we reach here, the whole first word is a prefix for all the other words
        return strs[0]
      
# Solution 2 (Horizontal Scanning) | O(S) time, O(1) space

# Solution 3 (Divide & Conquer) | O(S) time, O(m * logn) space  <- n strings, m is maxLen(all strings)

# Solution 4 (Binary Search) | O(S * logm) time, O(1) space
